pkg/apis/mysql/v1alpha1/mysqlcluster_defaults.go

```diff
@@ -0,0 +1,166 @@
+/*
+Copyright 2018 Pressinfra SRL
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	"strconv"
+
+	apiv1 "k8s.io/api/core/v1"
+	core "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/presslabs/mysql-operator/pkg/options"
+)
+
+// nolint: deadcode, varcheck
+const (
+	_        = iota // ignore first value by assigning to blank identifier
+	kb int64 = 1 << (10 * iota)
+	mb
+	gb
+)
+
+const (
+	resourceRequestCPU    = "200m"
+	resourceRequestMemory = "1Gi"
+
+	resourceStorage = "1Gi"
+
+	defaultMinAvailable = "50%"
+)
+
+// SetDefaults sets the defaults for Spec and Status
+func (c *MysqlCluster) SetDefaults(opt *options.Options) {
+	c.Spec.SetDefaults(opt, c)
+}
+
+// SetDefaults updates Spec defaults
+// nolint: gocyclo
```

Comment:
You should split this function into smaller functions, so that it passes gocyclo linting.

Reply:
Let's use set defaults from `pkg/wrappers/mysqlcluster`. I split the logic from `pkg/apis/mysql/v1alpha1` in `wrappers` package to separate the logic from data object. And when we need logic on an object we just wrap that object into a wrapper and all needed functions will be there. 

Doing so will be easier to import a clearer data structure into other projects. And the logic from wrappers  can depend on other packages that will not be imported in other projects.
 
@calind what do you think about this?

Reply:
I agree with this, except defaulting should stay near the type definition. See how code on kubernetes itself is organised. https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/core/v1/defaults.go